File: Imputation/lassosum/OTTERS_lassosum.R
#!/usr/bin/env Rscript

###################################################################
# Import packages needed
library(foreach)
library(bigstatsr)
library(data.table)
library(lassosum)

###############################################################
# parse input arguments
Sys.setlocale("LC_ALL", "C")
options(stringsAsFactors = F)

## Collect arguments
args <- commandArgs(TRUE)

## Default setting when no arguments passed
if(length(args) < 1) {
  args <- c("--help")
}

## Parse arguments (we expect the form --arg=value)
parseArgs <- function(x) strsplit(sub("^--", "", x), "=")
argsDF <- as.data.frame(do.call("rbind", parseArgs(args)))
argsL <- as.list(as.character(argsDF$V2))
names(argsL) <- argsDF$V1

## Check mandatory parameters
if (is.null(argsL$chr)) {
  cat('* Please specify the chromosome --chr\n')
  q(save="no")
} else if(is.null(argsL$gene_name)) {
  cat('* Please specify the path to the median sample size file using --medianN_path\n')
  q(save="no")
} else if(is.null(argsL$medianN)) {
  cat('* Please specify the path to the median sample size file using --medianN_path\n')
  q(save="no")
} else if (is.null(argsL$bim_file)) {
  cat('* Please specify the directory to the reference PLINK file --bim_dir\n')
  q(save="no")
} else if (is.null(argsL$sst_file)) {
  cat('* Please specify the path to the summary statistics with standardized beta using --sst_dir\n')
  q(save="no")
} else if (is.null(argsL$LDblocks)) {
  cat('* Please specify the name of LDblocks --LDblocks\n')
  q(save="no")
} else if (is.null(argsL$out_path)) {
  cat('* Please specify the output path\n')
  q(save="no")
} 

## Check optional parameters and assign default values
if (is.null(argsL$n_thread)){
  argsL$n_thread <- 1
}

print(argsL)

###############################################################
# time calculation
start_time <- Sys.time()

# Create the output file
gene_name=argsL$gene_name
chr = argsL$chr

# Specify the PLINK file of the reference panel
bfile <- argsL$bim_file
# Read the summary statistics of standardized beta in single variant test
# the standardized beta in single variant test = correlation
ss <- fread(argsL$sst_file)
cor <- ss$Beta
ss$SNPPos <- sapply(1:length(ss$SNP), function(i) strsplit(ss$SNP[i], "_")[[1]][2])
ss$Chrom <- chr

# train lassosum
out <- lassosum.pipeline(cor=cor,
                         chr=as.numeric(ss$Chrom),
                         pos=as.numeric(ss$SNPPos),
                         A1=ss$A1,
                         A2=ss$A2, # A2 is not required but advised
                         s = c(0.2, 0.5, 0.9, 1),
                         lambda = exp(seq(log(0.0001), log(0.1), length.out = 20)),
                         ref.bfile = bfile, # The reference panel dataset
                         test.bfile = bfile, # We don't have test data here
                         LDblocks = argsL$LDblocks,
                         exclude.ambiguous = F,
                         destandardize = F,
                         trace = 0)

# perform pseudovalidation
v <- pseudovalidate(out)
lassosum_out <- subset(out, s=v$best.s, lambda=v$best.lambda)

# save estimated beta
sumstats = lassosum_out$sumstats[, c("chr", "pos", "A1", "A2")]
beta = unlist(lassosum_out$beta)
results = data.frame(sumstats, TargetID = gene_name, ES = beta) 
results = results[, c("chr", "pos", "A1", "A2", "TargetID", "ES")]

write.table(results,
            argsL$out_path,
            quote = F,
            row.names= F,
            col.names= F,
            sep = "\t",
            append = T)

File: Imputation/lassosum/OTTERS_lassosum.py
#!/usr/bin/env python

"""
A TWAS method that infers posterior eQTL effect sizes under continuous shrinkage (CS) priors
using eQTL summary statistics and an external LD reference panel.

Usage:
python OTTERS_PRScs.py --OTTERS_dir=PATH_TO_OTTERS --in_dir=INPUR_DIR --out_dir=OUTPUT_DIR --chrom=CHROM
                [--window=WINDOW_SIZE --a=PARAM_A --b=PARAM_B --phi=PARAM_PHI --n_iter=MCMC_ITERATIONS --n_burnin=MCMC_BURNIN --thin=MCMC_THINNING_FACTOR --thread=THREAD --seed=SEED]

 - PATH_TO_OTTERS: The directory of OTTERS source code

 - INPUT_DIR:  Full path and the file name of the LD reference file.

 - OUTPUT_DIR: Output directory and output filename of the posterior effect size estimates.

 - CHROM: An integer for the chromosome of tested genes.

 - THREAD (optional): Number of simultaneous processes to use for parallel computation. Default is 1.

"""

import multiprocessing
import subprocess
import getopt
import sys
import os

import numpy as np
import pandas as pd

from time import time

############################################################
# time calculation
start_time = time()

############################################################


def parse_param():
    long_opts_list = ['OTTERS_dir=', 'in_dir=',
                      'out_dir=', 'chrom=', 'thread=', 'help']

    param_dict = {'OTTERS_dir': None, 'in_dir': None,
                  'out_dir': None, 'chrom': None,
                  'thread': 1}

    print('\n')

    if len(sys.argv) > 1:
        try:
            opts, args = getopt.getopt(sys.argv[1:], "h", long_opts_list)
        except:
            print('Option not recognized.')
            print('Use --help for usage information.\n')
            sys.exit(2)

        for opt, arg in opts:
            if opt == "-h" or opt == "--help":
                print(__doc__)
                sys.exit(0)
            elif opt == "--OTTERS_dir":
                param_dict['OTTERS_dir'] = arg
            elif opt == "--in_dir":
                param_dict['in_dir'] = arg
            elif opt == "--out_dir":
                param_dict['out_dir'] = arg
            elif opt == "--chrom":
                param_dict['chrom'] = str(arg)
            elif opt == "--thread":
                param_dict['thread'] = int(arg)
    else:
        print(__doc__)
        sys.exit(0)

    if param_dict['OTTERS_dir'] is None:
        print('* Please specify the directory to OTTERS --OTTERS_dir\n')
        sys.exit(2)
    elif param_dict['in_dir'] is None:
        print('* Please specify the full path to the sample size file using --anno_dir\n')
        sys.exit(2)
    elif param_dict['chrom'] is None:
        print('* Please specify the chromosome --chrom\n')
        sys.exit(2)
    elif param_dict['out_dir'] is None:
        print('* Please specify the output directory using --out_dir\n')
        sys.exit(2)

    for key in param_dict:
        print('--%s=%s' % (key, param_dict[key]))

    print('\n')
    return param_dict


############################################################

param_dict = parse_param()
sys.path.append(param_dict['OTTERS_dir'])
import OTTERSutils as ots

chrom = param_dict['chrom']
input_dir = param_dict['in_dir']
out_dir = param_dict['out_dir']
lassosum_path = os.path.join(param_dict['OTTERS_dir'], 'Imputation', 'lassosum', 'OTTERS_lassosum.R')

print('Reading sample size file.\n')
N_chunks = pd.read_csv(
    os.path.join(input_dir, 'medianN.txt'),
    sep='\t',
    header=0,
    chunksize=10000,
    iterator=True,
    dtype={'CHROM': object, 'TargetID': object, 'N': np.int32})

GeneN = pd.concat([x[x['CHROM'] == chrom] for x in N_chunks]).reset_index(drop=True)

if GeneN.empty:
    raise SystemExit('There are no valid sample size data for chromosome ' + chrom + '\n')

TargetID = GeneN.TargetID
n_targets = TargetID.size
N = GeneN.N

# create output file for PRS-CS
out_cols = ['CHROM', 'POS', 'A1', 'A2', 'TargetID', 'ES']
pd.DataFrame(columns=out_cols).to_csv(
    out_dir,
    sep='\t',
    index=None,
    header=True,
    mode='w')

# ############################################################

@ots.error_handler
def thread_process(num):
    print('Reading eQTL summary statistics data.')

    target = TargetID[num]
    print('num=' + str(num) + '\nTargetID=' + target)
    target_n = N[num]
    target_dir = os.path.join(input_dir, target)

    print("Start lassosum")

    try:
        lassosum_arg = ['Rscript ' + lassosum_path +
                        ' --gene_name=' + target +
                        ' --medianN=' + str(int(target_n)) +
                        ' --bim_file=' + target +
                        ' --sst_file=' + target + '_beta.txt' +
                        ' --out_path=' + out_dir +
                        ' --chr=' + str(chrom) +
                        ' --LDblocks=EUR.hg38']

        proc = subprocess.check_call(lassosum_arg,
                                     stdout=subprocess.PIPE,
                                     cwd=target_dir,
                                     shell=True,
                                     bufsize=1)
    except subprocess.CalledProcessError:
            print('SDPR failed for TargetID: ' + target)
            return None

    print('Done training lassosum for ' + str(num) + ':' + target)

############################################################
if __name__ == '__main__':
    print('Starting summary for ' + str(n_targets) + ' target genes.\n')
    pool = multiprocessing.Pool(param_dict['thread'])
    pool.imap(thread_process, [num for num in range(n_targets)])
    pool.close()
    pool.join()
    print('Done.')

############################################################
# time calculation
elapsed_sec = time()-start_time
elapsed_time = ots.format_elapsed_time(elapsed_sec)
print('Computation time (DD:HH:MM:SS): ' + elapsed_time)
